File: src/main/java/april_2025/EvalRPN.java
package april_2025;

import java.util.HashMap;
import java.util.Map;
import java.util.Stack;
import java.util.function.BiFunction;
import java.util.function.Function;

public class EvalRPN {

    // O(n) time | O(n) space
    public int evalRPN(String[] tokens) {
        Map<String, BiFunction<Integer, Integer, Integer>> mapping = new HashMap<>();
        mapping.put("+", (a, b) -> a + b);
        mapping.put("-", (a, b) -> a - b);
        mapping.put("*", (a, b) -> a * b);
        mapping.put("/", (a, b) -> a / b);

        Stack<Integer> stack = new Stack<>();
        for (String token : tokens) {
            if (mapping.containsKey(token)) {
                int b = stack.pop();
                int a = stack.pop();
                int calc = mapping.get(token).apply(a, b);
                stack.push(calc);
            } else {
                stack.push(Integer.valueOf(token));
            }
        }
        return stack.pop();
    }

}

File: python/eval_rpn.py
class Solution(object):
    # O(n) time | O(n) space
    def evalRPN(self, tokens):
        """
        :type tokens: List[str]
        :rtype: int
        """
        stack = []

        for token in tokens:
            if token in "+-*/":
                b = stack.pop()
                a = stack.pop()

                if token == '+':
                    stack.append(a + b)
                elif token == '-':
                    stack.append(a - b)
                elif token == '*':
                    stack.append(a * b)
                else:
                    stack.append(int(float(a) / b))

            else:
                stack.append(int(token))
        
        return stack.pop()

solution = Solution()
print(solution.evalRPN(["10","6","9","3","+","-11","*","/","*","17","+","5","+"]))